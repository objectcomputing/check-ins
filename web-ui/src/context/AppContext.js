import React, { useEffect, useReducer, useMemo } from "react";
import { createSelector } from 'reselect';
import { getCurrentUser, updateMember, getAllMembers } from "../api/member";
import { getCheckinByMemberId, createCheckin } from "../api/checkins";
import { BASE_API_URL } from "../api/api";
import axios from "axios";

export const MY_PROFILE_UPDATE = "@@check-ins/update_profile";
export const UPDATE_USER_BIO = "@@check-ins/update_bio";
export const UPDATE_CHECKINS = "@@check-ins/update_checkins";
export const UPDATE_INDEX = "@@check-ins/update_index";
export const UPDATE_TOAST = "@@check-ins/update_toast";
export const UPDATE_CURRENT_CHECKIN = "@@check-ins/update_current_checkin";
export const UPDATE_TEAMS = "@@check-ins/update_teams";
export const UPDATE_MEMBER_PROFILES = "@@check-ins/update_member_profiles";
export const UPDATE_TEAM_MEMBERS = "@@check-ins/update_team_members";
export const UPDATE_SELECTED_PROFILE = "@@check-ins/update_selected_profile";
const SET_CSRF = "@@check-ins/update_csrf";

const AppContext = React.createContext();

const reducer = (state, action) => {
  switch (action.type) {
    case MY_PROFILE_UPDATE:
      state.userProfile = action.payload;
      break;
    case UPDATE_USER_BIO:
      state.userProfile.memberProfile.bioText = action.payload;
      updateMember(state.userProfile.memberProfile);
      break;
    case UPDATE_CHECKINS:
      state.checkins = action.payload;
      //sort by date
      state.checkins.sort(function (a, b) {
        return new Date(...a.checkInDate) - new Date(...b.checkInDate);
      });
      state.currentCheckin = state.checkins[state.checkins.length - 1];
      break;
    case SET_CSRF:
      state.csrf = action.payload;
      break;
    case UPDATE_TOAST:
      state.toast = action.payload;
      break;
    case UPDATE_TEAMS:
      state.teams = action.payload;
      break;
    case UPDATE_MEMBER_PROFILES:
      state.memberProfiles = action.payload;
      break;
    case UPDATE_TEAM_MEMBERS:
      state.teamMembers = action.payload;
      break;
    case UPDATE_CURRENT_CHECKIN:
      state.currentCheckin = action.payload;
      break;
    case UPDATE_SELECTED_PROFILE:
      const { payload } = action;
      if (state.selectedProfile !== payload) {
        state.selectedProfile = payload;
        state.currentCheckin = payload ? payload.checkIn : {};
      }
      if (payload === undefined) {
        state.checkins = [];
      }
      break;
    default:
  }
  return { ...state };
};

const initialState = {
  checkins: [],
  csrf: undefined,
  currentCheckin: {},
  teams: [],
  memberProfiles: [],
  index: 0,
  selectedProfile: undefined,
  toast: {
    severity: "",
    toast: "",
  },
  userProfile: undefined,
};

const getCheckins = async (id, pdlId, date, dispatch, csrf) => {
  const res = await getCheckinByMemberId(id, csrf);
  let data =
    res.payload && res.payload.data && res.payload.status === 200 && !res.error
      ? res.payload.data
      : null;
  if (data && data.length > 0) {
    const allComplete = data.every((checkin) => checkin.completed === true);
    if (allComplete) {
      const prevCheckinDate = data[data.length - 1].checkInDate;
      if (pdlId) {
        const res = await createCheckin({
          teamMemberId: id,
          pdlId: pdlId,
          checkInDate: date(3, prevCheckinDate),
          completed: false,
        });
        const checkin =
          res.payload && res.payload.data && !res.error
            ? res.payload.data
            : null;
        data.push(checkin);
      }
    }
  } else if (data && data.length === 0) {
    if (pdlId) {
      const res = await createCheckin({
        teamMemberId: id,
        pdlId: pdlId,
        checkInDate: date(1),
        completed: false,
      });
      const checkin =
        res.payload && res.payload.data && !res.error ? res.payload.data : null;
      data = [checkin];
    }
  }
  //without this check you get infinite checkin calls
  if (data.length > 0) {
    dispatch({ type: UPDATE_CHECKINS, payload: data });
  }
};

const AppContextProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const memberProfile =
    state && state.userProfile && state.userProfile.memberProfile
      ? state.userProfile.memberProfile
      : undefined;
  const id = memberProfile ? memberProfile.id : undefined;
  const selectedProfile = state && state.selectedProfile;
  const selectedId = selectedProfile ? selectedProfile.id : undefined;


  const { csrf } = state;

  useEffect(() => {
    const getCsrf = async () => {
      if (!csrf) {
        const res = await axios({
          url: `${BASE_API_URL}/csrf/cookie`,
          responseType: "text",
          withCredentials: true,
        });
        if (res && res.data) {
          dispatch({ type: SET_CSRF, payload: res.data });
        }
      }
    };
    getCsrf();
  }, [csrf]);

  useEffect(() => {
    const updateUserProfile = async () => {
      let res = await getCurrentUser(csrf);
      let profile =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profile) {
        dispatch({ type: MY_PROFILE_UPDATE, payload: profile });
      }
    };
    if (csrf) {
      updateUserProfile();
    }
  }, [csrf]);

  const date = (months, prevCheckinDate) => {
    let currentMonth = new Date().getMonth() + 1;
    let newDate = prevCheckinDate ? new Date(...prevCheckinDate) : new Date();
    if (prevCheckinDate) {
      let prevCheckinMonth = prevCheckinDate.getMonth() + 1;
      newDate.setMonth(
        newDate.getMonth() + currentMonth - prevCheckinMonth >= 3 ? 1 : months
      );
    } else {
      newDate.setMonth(newDate.getMonth() + months);
    }
    const year = newDate.getFullYear();
    const month = newDate.getMonth() + 1;
    const day = newDate.getDate();
    const hours = newDate.getHours();
    const minutes = newDate.getMinutes();
    const dateTimeArray = [year, month, day, hours, minutes, 0];
    return dateTimeArray;
  };

  useEffect(() => {
    async function getMemberProfiles() {
      let res = await getAllMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_MEMBER_PROFILES, payload: profiles });
      }
    }
    if (csrf) {
      getMemberProfiles();
    }
  }, [csrf]);

  useEffect(() => {
    if (selectedId && csrf) {
      getCheckins(selectedId, id, date, dispatch, csrf);
    }
  }, [csrf, selectedId, id]);

  const value = useMemo(() => {
    return { state, dispatch };
  }, [state]);
  return (
    <AppContext.Provider value={props.value || value}>
      {props.children}
    </AppContext.Provider>
  );
};

export const selectMemberProfiles = state => state.memberProfiles;
export const selectTeamMembers = state => state.teamMembers;

export const selectProfileMap = createSelector(
  selectMemberProfiles,
  (memberProfiles) => {
    if (memberProfiles && memberProfiles.length) {
      memberProfiles = memberProfiles.reduce((mappedById, profile) => {
        mappedById[profile.id] = profile;
        return mappedById;
      }, {});
    }
    return memberProfiles;
  }
);

export const selectMembersByTeamId = id => createSelector(
  selectTeamMembers,
  (teamMembers) => {
    let members = [];
    if (teamMembers && teamMembers.length) {
      members = teamMembers.filter((member) => member.teamid === id);
    }
    return members;
  }
);

export const selectMemberProfilesByTeamId = id => createSelector(
  selectMembersByTeamId(id),
  selectProfileMap,
  (members, profileMap) => members.map((member) => {
      return { ...profileMap[member.memberid], ...member };
    })
);

export { AppContext, AppContextProvider };
