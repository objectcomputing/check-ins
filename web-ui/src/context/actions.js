export const MY_PROFILE_UPDATE = "@@check-ins/update_profile";
export const UPDATE_USER_BIO = "@@check-ins/update_bio";
export const UPDATE_CHECKINS = "@@check-ins/update_checkins";
export const UPDATE_CHECKIN = "@@check-ins/update_checkin";
export const ADD_CHECKIN = "@@check-ins/add_checkin";
export const UPDATE_TOAST = "@@check-ins/update_toast";
export const UPDATE_MEMBER_SKILLS = "@@check-ins/update_member_skills";
export const DELETE_MEMBER_SKILL = "@@check-ins/delete_member_skill";
export const ADD_MEMBER_SKILL = "@@check-ins/add_member_skill";
export const UPDATE_TEAMS = "@@check-ins/update_teams";
export const UPDATE_MEMBER_PROFILES = "@@check-ins/update_member_profiles";
export const UPDATE_TEAM_MEMBERS = "@@check-ins/update_team_members";
export const ADD_SKILL = "@@check-ins/add-skill";
export const DELETE_SKILL = "@@check-ins/delete-skill";
export const UPDATE_SKILL = "@@check-ins/update_skill";
export const UPDATE_SKILLS = "@@check-ins/update_skills";
export const ADD_TEAM = "@@check-ins/add_team";
export const SET_CSRF = "@@check-ins/update_csrf";
export const SET_ROLES = "@@check-ins/set_roles";
