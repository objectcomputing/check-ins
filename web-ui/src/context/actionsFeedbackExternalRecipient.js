export const SET_CSRF = '@@check-ins/update_csrf';
export const UPDATE_TOAST = '@@check-ins/update_toast';

