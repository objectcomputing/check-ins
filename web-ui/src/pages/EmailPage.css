.email-page .stepper-button-container {
    margin-top: 1rem;
    text-align: right;
}

.email-page .stepper-button {
    margin-left: 1rem;
}

.email-page .current-step-content-card {
    margin-top: 1rem;
}

.email-page .current-step-content-card .file-preview-container {
    margin-top: 2rem;
}

.email-page .current-step-content-card .file-preview-container .file-preview {
    background-color: #282c34;
    color: white;
    width: 100%;
    border: none;
    outline: none;
    resize: vertical;
    min-height: 70px;
}

.email-page .current-step-content-card .send-test-email-container {
    display: flex;
    flex-direction: row;
    align-items: flex-start;
    margin-top: 2rem;
}

.email-page .current-step-content-card .send-test-email-input {
    width: 400px;
}

.email-page .current-step-content-card .send-test-email-button {
    height: 56px;
    border-bottom-left-radius: 0;
    border-top-left-radius: 0;
}

.send-email-to-all-confirmation-dialog {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    max-width: 700px;
    padding:0.5rem;
}