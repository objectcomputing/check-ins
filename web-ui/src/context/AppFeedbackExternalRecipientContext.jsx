import React, { useEffect, useReducer, useMemo } from 'react';
import {reducerFeedbackExternalRecipient, initialState} from "./reducerFeedbackExternalRecilpients.js";
import {BASE_API_URL} from "../api/apiFeedbackExternalRecipient.js";
import {SET_CSRF} from "./actionsFeedbackExternalRecipient.js";

const AppFeedbackExternalRecipientContext = React.createContext();

function getSessionCookieValue(name) {
  const cookies = document?.cookie?.split(';');
  for (let i = 0; i < cookies.length; i++) {
    const cookie = cookies[i].trim();
    if (cookie.startsWith(name + '=')) {
      return decodeURIComponent(cookie.substring(name.length + 1));
    }
  }
  return null;
}

const AppFeedbackExternalRecipientContextProvider = props => {
  console.log("AppFeedbackExternalRecipientContextProvider props: ", props);
  const [state, dispatch] = useReducer(
      reducerFeedbackExternalRecipient,
    props?.value?.state || initialState
  );

  const {
    csrf,
  } = state;
  const url = `${BASE_API_URL}/csrf/cookie`;

  useEffect(() => {

    console.log("AppFeedbackExternalRecipientContextProvider useEffect, csrf");

    const getCsrf = async () => {
      if (!csrf) {
        const payload = getSessionCookieValue('_csrf');
        if (payload) {
          dispatch({ type: SET_CSRF, payload });
        } else {
          const res = await fetch(url, {
            responseType: 'text',
            credentials: 'include'
          });
          if (res && res.ok) {
            dispatch({ type: SET_CSRF, payload: await res.text() });
          }
        }
      }
    };
    getCsrf();
  }, [csrf]);

  const value = useMemo(() => {
    console.log("AppFeedbackExternalRecipientContextProvider useMemo, state: ", state);
    console.log("AppFeedbackExternalRecipientContextProvider useMemo, dispatch: ", dispatch);
    return { state, dispatch };
  }, [state]);

  return (
    <AppFeedbackExternalRecipientContext.Provider value={props.value || value}>
      {props.children}
    </AppFeedbackExternalRecipientContext.Provider>
  );
};

export { AppFeedbackExternalRecipientContext, AppFeedbackExternalRecipientContextProvider };
