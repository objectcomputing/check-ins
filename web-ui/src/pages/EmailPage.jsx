import React, {useContext, useEffect, useState} from "react";
import {
  Alert,
  Button,
  Card,
  CardActions,
  CardContent,
  CardHeader,
  Modal,
  Step,
  StepLabel,
  Stepper,
  TextareaAutosize,
  TextField,
  Typography,
} from "@mui/material";
import {styled} from "@mui/material/styles";
import LeftArrowIcon from "@mui/icons-material/KeyboardArrowLeft";
import RightArrowIcon from "@mui/icons-material/KeyboardArrowRight";
import UploadFileIcon from "@mui/icons-material/UploadFile";
import SendIcon from "@mui/icons-material/Send";
import {AppContext} from "../context/AppContext";
import mjml2html from "mjml-browser";
import ReactHtmlParser from "react-html-parser";
import {UPDATE_TOAST} from "../context/actions";
import {sendEmail} from "../api/notifications";

import "./EmailPage.css";


const Root = styled("div")({
  margin: "2rem"
});

const UploadFileStep = ({ emailFile, emailContents, onEmailFileChange, onEmailContentsChange }) => {

  const handleFileUpload = (event) => {
    if (event.target.files && event.target.files[0]) {
      const fileReader = new FileReader();
      fileReader.onload = (e) => {
        const content = e.target.result.toString();
        onEmailContentsChange(content);
      }

      const file = event.target.files[0];
      onEmailFileChange(file);
      fileReader.readAsText(file);
    }
  }

  return (
    <>
      <Typography variant="body1">Select a MJML file to render the email. The file must have a .mjml extension.</Typography>
      <Button variant="contained" component="label" startIcon={<UploadFileIcon/>} disableElevation>
        Choose File
        <input
          type="file"
          accept=".mjml"
          hidden
          onChange={handleFileUpload}
        />
      </Button>
      <div className="file-preview-container">
        <Typography variant="body1" fontWeight={emailFile ? "bold" : "normal"}>{emailFile ? emailFile.name : "No file uploaded"}</Typography>
        {emailContents &&
          <TextareaAutosize
            className="file-preview"
            style={{overflow: "auto"}}
            value={emailContents}
            minRows={3}
            maxRows={emailContents.split("\n").length}
            readOnly/>
        }
      </div>
    </>
  );
}

const PreviewEmailStep = ({ emailContents, emailSubjectError, emailSubject, onSubjectChange }) => {

  if (!emailContents) {
    return (
      <div className="missing-preview-message">
        <Typography variant="h6">Preview not available</Typography>
      </div>
    );
  }

  const { html } = mjml2html(emailContents);
  const emailPreview = ReactHtmlParser(html);

  return (
    <>
      <div className="email-subject-container">
        <Typography
          variant="body1"
          fontWeight="bold"
          style={{marginRight: "1rem", marginTop: "6px"}}>
          Subject:
        </Typography>
        <TextField
          fullWidth
          error={emailSubjectError}
          helperText={emailSubjectError ? "Email is missing subject" : ""}
          value={emailSubject}
          onChange={(event) => {
            onSubjectChange(event.target.value);
          }}
        />
      </div>
      {emailPreview}
    </>
  );
}

const SendEmailStep = ({ testEmail, onTestEmailChange, onSendTestEmail }) => {
  const [emailError, setEmailError] = useState(false)

  const handleSendButtonClick = () => {
    let regEmail = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}])|(([a-zA-Z\-\d]+\.)+[a-zA-Z]{2,}))$/
    if (!regEmail.test(testEmail)) {
      setEmailError(true);
    } else {
      onSendTestEmail(testEmail);
    }
  }

  return (
    <>
      <Typography variant="body1">To test this email and see how it will appear in an inbox, you can send it to an email address you have access to. If you are satisfied with the email, click the send button at the bottom of the page to send to everyone.</Typography>
      <div className="send-test-email-container">
        <TextField
          className="send-test-email-input"
          label="Email address"
          placeholder="example@objectcomputing.com"
          variant="outlined"
          value={testEmail}
          error={emailError}
          helperText={emailError ? "Invalid email address" : ""}
          onChange={(event) => {
            onTestEmailChange(event.target.value);
            setEmailError(false);
          }}
          InputProps={{
            style: {borderTopRightRadius: 0, borderBottomRightRadius: 0}
          }}
        />
        <Button
          className="send-test-email-button"
          variant="contained"
          disableElevation
          endIcon={<SendIcon/>}
          disabled={testEmail.trim().length === 0 || emailError}
          onClick={handleSendButtonClick}
        >
          Send Test Email
        </Button>
      </div>
    </>
  );
}

const EmailPage = () => {
  const { state } = useContext(AppContext);
  const { memberProfiles, csrf } = state;
  const [currentStep, setCurrentStep] = useState(0);
  const [emailFile, setEmailFile] = useState(null);
  const [emailContents, setEmailContents] = useState("");
  const [emailSubject, setEmailSubject] = useState("");
  const [emailSubjectError, setEmailSubjectError] = useState(false);
  const [testEmail, setTestEmail] = useState("");
  const [testEmailSent, setTestEmailSent] = useState(false);
  const [confirmationDialogOpen, setConfirmationDialogOpen] = useState(false);
  const [activeMembers, setActiveMembers] = useState([]);
  const steps = ["Upload File", "Preview Email", "Send Email"];

  useEffect(() => {
    const unterminatedMembers = memberProfiles.filter(member => member.terminationDate === null);
    setActiveMembers(unterminatedMembers);
  }, [memberProfiles]);

  const sendTestEmail = () => {

    if (!emailSubject.trim() || !emailContents || !csrf) {
      return;
    }

    sendEmail(emailSubject, emailContents, testEmail, csrf).then(res => {
      let toastMessage, toastStatus;
      if (res && res.payload && res.payload.status === 200 && !res.error) {
        setTestEmailSent(true);
        toastStatus = "success";
        toastMessage = `Sent a test email to ${testEmail}`;
      } else {
        toastStatus = "error";
        toastMessage = `Failed to send test email to ${testEmail}`;
      }
      window.snackDispatch({
        type: UPDATE_TOAST,
        payload: {
          severity: toastStatus,
          toast: toastMessage
        }
      });
    });
  }

  const sendEmailToAllMembers = () => {
    setConfirmationDialogOpen(false);
    window.snackDispatch({
      type: UPDATE_TOAST,
      payload: {
        severity: "success",
        toast: `Sent email to ${activeMembers.length} members`
      }
    });
  }

  const handleNextClick = () => {
    switch (currentStep) {
      case 0:
        setCurrentStep(currentStep + 1);
        break;
      case 1:
        if (emailSubject.trim().length > 0) {
          setCurrentStep(currentStep + 1);
        } else {
          setEmailSubjectError(true);
        }
        break;
      case 2:
        setConfirmationDialogOpen(true);
        break;
      default:
        console.warn(`Invalid step in stepper: ${currentStep}`);
    }
  }

  return (
    <Root className="email-page">
      <Stepper activeStep={currentStep}>
        {steps.map((step) => (
          <Step key={step}>
            <StepLabel>{step}</StepLabel>
          </Step>
        ))}
      </Stepper>
      <Card className="current-step-content-card">
        <CardHeader title={steps[currentStep]}/>
        <CardContent>
          {currentStep === 0 && (
            <UploadFileStep
              emailFile={emailFile}
              emailContents={emailContents}
              onEmailFileChange={(file) => setEmailFile(file)}
              onEmailContentsChange={(content) => setEmailContents(content)}
            />
          )}
          {currentStep === 1 && (
            <PreviewEmailStep
              emailContents={emailContents}
              emailSubjectError={emailSubjectError}
              emailSubject={emailSubject}
              onSubjectChange={(subject) => {
                setEmailSubject(subject);
                setEmailSubjectError(false);
              }}
            />
          )}
          {currentStep === 2 && (
            <SendEmailStep
              testEmail={testEmail}
              onTestEmailChange={(address) => setTestEmail(address)}
              onSendTestEmail={sendTestEmail}/>
          )}
        </CardContent>
      </Card>
      <div className="stepper-button-container">
        <Button
          className="stepper-button"
          variant="outlined"
          startIcon={<LeftArrowIcon/>}
          disabled={currentStep === 0}
          onClick={() => currentStep > 0 ? setCurrentStep(currentStep - 1) : null}>
          Back
        </Button>
        <Button
          className="stepper-button"
          variant="contained"
          color={currentStep === steps.length - 1 ? "success" : "primary"}
          endIcon={currentStep === steps.length - 1 ? <SendIcon/> : <RightArrowIcon/>}
          disabled={currentStep === steps.length - 1 && !emailFile}
          onClick={handleNextClick}>
          {currentStep === steps.length - 1 ? "Send" : "Next"}
        </Button>
      </div>
      <Modal open={confirmationDialogOpen}>
        <Card className="send-email-to-all-confirmation-dialog">
          <CardHeader title={<Typography variant="h5" fontWeight="bold">Send Email</Typography>}/>
          <CardContent>
            <Typography variant="body1">
              You are about to send the email <b>{emailFile?.name}</b> to everyone in Check-Ins ({activeMembers.length} members). Are you sure?
            </Typography>
            {!testEmailSent &&
              <Alert severity="warning" style={{marginTop: "1rem"}}>
                Caution: You have not sent a test email to check the email formatting.
              </Alert>
            }
          </CardContent>
          <CardActions>
            <Button style={{ color: "gray" }} onClick={() => setConfirmationDialogOpen(false)}>
              Cancel
            </Button>
            <Button color="primary" onClick={emailFile && sendEmailToAllMembers} disabled={!emailFile}>
              Send
            </Button>
          </CardActions>
        </Card>
      </Modal>
    </Root>
  );
};

export default EmailPage;