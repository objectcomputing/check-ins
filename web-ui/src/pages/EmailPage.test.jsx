import React from 'react';
import EmailPage from './EmailPage';
import { AppContextProvider } from '../context/AppContext';
import { BrowserRouter } from 'react-router-dom';

window.scrollTo = vi.fn();

const userStateWithPermission = {
  state: {
    userProfile: {
      name: 'Mitch Hedberg',
      role: ['MEMBER'],
      permissions: [{ permission: 'CAN_SEND_EMAIL' }],
    },
    terminatedMembers: [],
  }
};

it('renders correctly', () => {
  snapshot(
    <AppContextProvider value={userStateWithPermission}>
      <BrowserRouter>
        <EmailPage />
      </BrowserRouter>
    </AppContextProvider>
  );
});

it('renders an error if user does not have appropriate permission', () => {
  snapshot(
    <AppContextProvider>
      <BrowserRouter>
        <EmailPage />
      </BrowserRouter>
    </AppContextProvider>
  );
});
