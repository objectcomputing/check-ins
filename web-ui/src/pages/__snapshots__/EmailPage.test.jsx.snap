// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders an error if user does not have appropriate permission 1`] = `
<div>
  <h3>
    You do not have permission to view this page.
  </h3>
</div>
`;

exports[`renders correctly 1`] = `
<div>
  <div
    class="email-page css-175t7kh"
  >
    <div
      class="MuiStepper-root MuiStepper-horizontal css-m5vj9m-MuiStepper-root"
    >
      <div
        class="MuiStep-root MuiStep-horizontal css-1bw0nnu-MuiStep-root"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal css-ascpo7-MuiStepLabel-root"
        >
          <span
            class="MuiStepLabel-iconContainer Mui-active css-vnkopk-MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root Mui-active css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                dominant-baseline="central"
                text-anchor="middle"
                x="12"
                y="12"
              >
                1
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer css-1vyamtt-MuiStepLabel-labelContainer"
          >
            <span
              class="MuiStepLabel-label Mui-active css-1hv8oq8-MuiStepLabel-label"
            >
              Choose Email Format
            </span>
          </span>
        </span>
      </div>
      <div
        class="MuiStepConnector-root MuiStepConnector-horizontal Mui-disabled css-j5w0w9-MuiStepConnector-root"
      >
        <span
          class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
        />
      </div>
      <div
        class="MuiStep-root MuiStep-horizontal css-1bw0nnu-MuiStep-root"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled css-ascpo7-MuiStepLabel-root"
        >
          <span
            class="MuiStepLabel-iconContainer Mui-disabled css-vnkopk-MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                dominant-baseline="central"
                text-anchor="middle"
                x="12"
                y="12"
              >
                2
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer css-1vyamtt-MuiStepLabel-labelContainer"
          >
            <span
              class="MuiStepLabel-label Mui-disabled css-1hv8oq8-MuiStepLabel-label"
            >
              Compose Email
            </span>
          </span>
        </span>
      </div>
      <div
        class="MuiStepConnector-root MuiStepConnector-horizontal Mui-disabled css-j5w0w9-MuiStepConnector-root"
      >
        <span
          class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
        />
      </div>
      <div
        class="MuiStep-root MuiStep-horizontal css-1bw0nnu-MuiStep-root"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled css-ascpo7-MuiStepLabel-root"
        >
          <span
            class="MuiStepLabel-iconContainer Mui-disabled css-vnkopk-MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                dominant-baseline="central"
                text-anchor="middle"
                x="12"
                y="12"
              >
                3
              </text>
            </svg>
          </span>
          <span
            class="MuiStepLabel-labelContainer css-1vyamtt-MuiStepLabel-labelContainer"
          >
            <span
              class="MuiStepLabel-label Mui-disabled css-1hv8oq8-MuiStepLabel-label"
            >
              Select Recipients
            </span>
          </span>
        </span>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root current-step-content-card css-bhp9pd-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
          >
            Choose Email Format
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          class="email-format-container"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary email-format-button css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <div
              class="email-format-button-content"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bz1pui-MuiSvgIcon-root"
                data-testid="UploadFileIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8zm4 18H6V4h7v5h5zM8 15.01l1.41 1.41L11 14.84V19h2v-4.16l1.59 1.59L16 15.01 12.01 11z"
                />
              </svg>
              <h6
                class="MuiTypography-root MuiTypography-h6 css-tjvwag-MuiTypography-root"
              >
                MJML File
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
              >
                Create an email with a custom format using MJML
              </p>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary email-format-button css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <div
              class="email-format-button-content"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bz1pui-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                />
              </svg>
              <h6
                class="MuiTypography-root MuiTypography-h6 css-tjvwag-MuiTypography-root"
              >
                Text
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
              >
                Write a simple email with no formatting
              </p>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="stepper-button-container"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary stepper-button css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6z"
            />
          </svg>
        </span>
        Back
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary stepper-button css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Next
        <span
          class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
